### PR TITLE
Improvements to Espresso tests to make tests more stable

### PR DESCRIPTION
**Pull request by seadowg**
I'll leave some comments inline to explain changes. I've also put up a [document](https://github.com/seadowg/espresso-tips) of "Espresso tips" related to these changes up. I started this a while ago in my own notebook but thought it would be good to start collecting some of my learnings publicly as Android's documentation is falling behind somewhat.

#### What has been done to verify that this works as intended?

Ran this branch on test lab **a lot** of times. Haven't seen any fails in the current change set (this doesn't mean they aren't there though).

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Only changes to tests so this can be merged without QA.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/build.gradle [hunk 1]
```diff
@@ -327,6 +327,7 @@ dependencies {
     androidTestImplementation "org.mockito:mockito-android:3.3.3"
     androidTestImplementation 'androidx.test.ext:junit:1.1.1'
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.2.0'
+    androidTestImplementation "androidx.arch.core:core-testing:2.1.0"
```
Comment (seadowg) [inline collect_app/build.gradle:330]: We need this for `CountingTaskExecutorRule`

File: collect_app/src/androidTest/java/org/odk/collect/android/regression/AdminSettingsTest.java [hunk 2]
```diff
@@ -1,20 +1,38 @@
...
+
+    @Rule
+    public RuleChain copyFormChain = RuleChain
+            .outerRule(GrantPermissionRule.grant(
+                    Manifest.permission.READ_EXTERNAL_STORAGE,
+                    Manifest.permission.WRITE_EXTERNAL_STORAGE,
+                    Manifest.permission.READ_PHONE_STATE
+            ))
+            .around(new ResetStateRule())
+            .around(rule);
```
Comment (seadowg) [inline collect_app/src/androidTest/java/org/odk/collect/android/regression/AdminSettingsTest.java:30]: This test needed permissions but as we've seen before it's important to chain the rules so permissions are definitely granted **before** any Activities launch.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/CountingTaskExecutorIdlingResource.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package org.odk.collect.android.support;
+
+import androidx.arch.core.executor.testing.CountingTaskExecutorRule;
+import androidx.test.espresso.IdlingResource;
+
+public class CountingTaskExecutorIdlingResource implements IdlingResource {
```
Comment (seadowg) [inline collect_app/src/androidTest/java/org/odk/collect/android/support/CountingTaskExecutorIdlingResource.java:6]: There is some good docs on `IdlingResource` [here](https://developer.android.com/training/testing/espresso/idling-resource). It seems Espresso is moving towards needing them more and more so it doesn't have to integrate with all the different scheduler mechanisms in Android (`Handler`, `ArchTaskExectuor`, coroutines and `WorkManager` to name a few).

File: collect_app/src/androidTest/java/org/odk/collect/android/support/pages/FormEntryPage.java [hunk 4]
```diff
@@ -54,6 +54,15 @@ public FormEntryPage swipeToNextQuestion(int repetitions) {
         return this;
     }
 
+    public FormEntryPage swipeToNextRepeat(String repeatLabel, int repeatNumber) {
+        tryAgainOnFail(() -> {
+            onView(withId(R.id.questionholder)).perform(swipeLeft());
```
Comment (seadowg) [inline collect_app/src/androidTest/java/org/odk/collect/android/support/pages/FormEntryPage.java:59]: We might need to move all of our `swipeToX` helpers to this kind of style. It's definitely safer for us to have assertions with actions like gestures that can be performed on any screen.

File: collect_app/src/androidTest/java/org/odk/collect/android/support/CountingTaskExecutorIdlingResource.java (hunk 3 above)
Comment (lognaturel) [reply collect_app/src/androidTest/java/org/odk/collect/android/support/CountingTaskExecutorIdlingResource.java:6]: Did the person on the issue you filed about `WorkManager` and `Espresso` ever get back to you and was this the answer?

**Review by lognaturel — APPROVED** [review]
I've had failures locally because some tests (draw widget, maybe) change the orientation to landscape and then other tests that go after it and are executed in landscape don't scroll to the desired UI element before clicking. That may be another class of issues to address at some point.

**Comment by seadowg** [comment]
@lognaturel That discussion is here: https://github.com/android/android-test/issues/462. Nope. No reply. I did some digging to find this stuff. I've been avoiding writing my own documentation for Espresso but I'm starting to give in!

**Pull request merged**